base/range.jl

```diff
@@ -871,13 +871,22 @@ end
 
 median(r::Range{<:Real}) = mean(r)
 
-function in(x, r::Range)
-    n = step(r) == 0 ? 1 : round(Integer,(x-first(r))/step(r))+1
-    n >= 1 && n <= length(r) && r[n] == x
+function _in_range(x, r::Range)
+    if step(r) == 0
+        return !isempty(r) && first(r) == x
+    else
+        n = round(Integer, (x - first(r)) / step(r)) + 1
+        return n >= 1 && n <= length(r) && r[n] == x
+    end
 end
+in(x::Real, r::Range{<:Real}) = _in_range(x, r)
+# This method needs to be defined separately since -(::T, ::T) can be implemented
+# even if -(::T, ::Real) is not
+in(x::T, r::Range{T}) where {T} = _in_range(x, r)
 
 in(x::Integer, r::AbstractUnitRange{<:Integer}) = (first(r) <= x) & (x <= last(r))
-in(x, r::Range{T}) where {T<:Integer} =
```

Comment:
Not that I know of. FWIW, `isinteger` is only defined for `Real`, `Complex` and `Irrational` in Base. `Number` is too broad for `<=`. This sounds like a typical case where a trait would be useful (once we have traits).

BTW, when you say:
> That seems considerably more expensive for inputs that are not Real or Complex.

To be precise, `in` will still be as fast as before for any arbitrary type. What will be slow is calling `in` with an element of a custom non-`Real` type different from the range's eltype. This is probably not very common.